File: scripts/update_readme.py
#!/usr/bin/env python3
"""
update_readme.py — rewrites dynamic zones in README.md using GitHub API data.

Dynamic zones are delimited by HTML comment markers:
  <!-- ZONE_START --> ... <!-- ZONE_END -->

Zones updated:
  - PROJECTS: top public repos ranked by stars + recency (no forks, no profile repo)
  - OSS:      recent external PRs from public events
  - STATS:    github-readme-stats image (static URL, auto-refreshes on render)

Requires:
  - GH_TOKEN env var (PAT with repo + read:user scope)
"""

import os
import re
import sys
from datetime import datetime, timezone
from urllib.request import Request, urlopen
from urllib.error import URLError
import json

GITHUB_USERNAME = "jeziellopes"
README_PATH = os.path.join(os.path.dirname(__file__), "..", "README.md")
API_BASE = "https://api.github.com"

TOP_PROJECTS_COUNT = 4
OSS_CONTRIBUTIONS_COUNT = 5


def gh_get(path: str, token: str) -> list | dict:
    url = f"{API_BASE}{path}"
    req = Request(url, headers={
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github+json",
        "X-GitHub-Api-Version": "2022-11-28",
    })
    with urlopen(req) as resp:
        return json.loads(resp.read())


def score_repo(repo: dict) -> float:
    """Rank repos by stars (primary) and recency (secondary)."""
    stars = repo.get("stargazers_count", 0)
    updated = repo.get("updated_at", "1970-01-01T00:00:00Z")
    days_ago = (datetime.now(timezone.utc) - datetime.fromisoformat(updated.replace("Z", "+00:00"))).days
    recency_score = max(0, 365 - days_ago) / 365  # 0–1 scale
    return stars * 10 + recency_score


def build_projects_section(token: str) -> str:
    repos = gh_get(f"/users/{GITHUB_USERNAME}/repos?sort=updated&per_page=100&type=owner", token)
    public = [
        r for r in repos
        if not r["private"]
        and not r["fork"]
        and r["name"] != GITHUB_USERNAME
    ]
    top = sorted(public, key=score_repo, reverse=True)[:TOP_PROJECTS_COUNT]

    rows = []
    for r in top:
        name = r["name"]
        url = r["html_url"]
        desc = (r.get("description") or "").replace("|", "\\|")
        stars = r.get("stargazers_count", 0)
        forks = r.get("forks_count", 0)
        rows.append(f"| [{name}]({url}) | {desc} | ⭐ {stars} | 🍴 {forks} |")

    lines = [
        "## 🚀 Featured Projects",
        "",
        "| Project | Description | Stars | Forks |",
        "|---------|-------------|-------|-------|",
        *rows,
    ]
    return "\n".join(lines)


def build_oss_section(token: str) -> str:
    events = gh_get(f"/users/{GITHUB_USERNAME}/events?per_page=100", token)
    contributions = []
    seen = set()

    for event in events:
        if event.get("type") != "PullRequestEvent":
            continue
        repo_name = event.get("repo", {}).get("name", "")
        owner = repo_name.split("/")[0] if "/" in repo_name else ""
        if owner == GITHUB_USERNAME:
            continue  # skip own repos

        pr_payload = event.get("payload", {}).get("pull_request", {})
        pr_api_url = pr_payload.get("url", "")
        pr_number = pr_payload.get("number")

        if not pr_api_url or pr_api_url in seen:
            continue
        seen.add(pr_api_url)

        # The events API returns a minimal PR object — fetch the full details
        try:
            pr = gh_get(pr_api_url.replace(API_BASE, ""), token)
        except Exception:
            continue

        pr_url = pr.get("html_url", f"https://github.com/{repo_name}/pull/{pr_number}")
        pr_title = pr.get("title") or repo_name
        merged = pr.get("merged_at")
        pr_state = pr.get("state", "open")

        if merged:
            status = "✅ Merged"
        elif pr_state == "open":
            status = "🔄 Open"
        else:
            status = "❌ Closed"

        repo_url = f"https://github.com/{repo_name}"
        contributions.append((pr_title, pr_url, repo_name, repo_url, status))
        if len(contributions) >= OSS_CONTRIBUTIONS_COUNT:
            break

    if not contributions:
        return "## 🤝 OSS Contributions\n\n_No recent external contributions found._"

    rows = [
        f"| [{title}]({pr_url}) | [{repo}]({repo_url}) | {status} |"
        for title, pr_url, repo, repo_url, status in contributions
    ]
    lines = [
        "## 🤝 OSS Contributions",
        "",
        "| PR | Repository | Status |",
        "|----|-----------|--------|",
        *rows,
    ]
    return "\n".join(lines)


def build_stats_section() -> str:
    url = (
        f"https://github-readme-stats.vercel.app/api"
        f"?username={GITHUB_USERNAME}"
        f"&show_icons=true"
        f"&count_private=true"
        f"&hide_border=true"
        f"&theme=dark"
    )
    return (
        "## 📊 GitHub Stats\n\n"
        f'<p>\n  <img src="{url}" />\n</p>'
    )


def rewrite_zone(content: str, zone: str, new_body: str) -> str:
    pattern = re.compile(
        rf"(<!-- {zone}_START -->)\n.*?\n(<!-- {zone}_END -->)",
        re.DOTALL,
    )
    replacement = rf"\1\n{new_body}\n\2"
    updated, count = pattern.subn(replacement, content)
    if count == 0:
        print(f"WARNING: zone {zone} not found in README", file=sys.stderr)
    return updated


def main() -> None:
    token = os.environ.get("GH_TOKEN", "")
    if not token:
        print("ERROR: GH_TOKEN environment variable is not set.", file=sys.stderr)
        sys.exit(1)

    readme_path = os.path.abspath(README_PATH)
    with open(readme_path, encoding="utf-8") as f:
        original = f.read()

    try:
        projects = build_projects_section(token)
        oss = build_oss_section(token)
        stats = build_stats_section()
    except URLError as e:
        print(f"ERROR: GitHub API request failed: {e}", file=sys.stderr)
        sys.exit(1)

    updated = original
    updated = rewrite_zone(updated, "PROJECTS", projects)
    updated = rewrite_zone(updated, "OSS", oss)
    updated = rewrite_zone(updated, "STATS", stats)

    if updated == original:
        print("README.md is already up to date.")
        return

    with open(readme_path, "w", encoding="utf-8") as f:
        f.write(updated)
    print("README.md updated.")


if __name__ == "__main__":
    main()

File: scripts/update_readme.go
// update_readme.go — rewrites dynamic zones in README.md using GitHub API data.
//
// Dynamic zones are delimited by HTML comment markers:
//
//	<!-- ZONE_START --> ... <!-- ZONE_END -->
//
// Zones updated:
//   - PROJECTS: top public repos ranked by stars + recency (no forks, no profile repo)
//   - OSS:      recent external PRs from public events (private repos filtered out)
//
// Usage:
//
//	GH_TOKEN=<pat> go run scripts/update_readme.go
//
// Required:
//
//	GH_TOKEN env var — Personal Access Token with scopes: repo, read:user

package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"sort"
	"strings"
	"time"
)

const (
	githubUsername    = "jeziellopes"
	apiBase           = "https://api.github.com"
	topProjectsCount  = 4
	ossContribCount   = 5
)

var readmePath = func() string {
	_, file, _, _ := runtime.Caller(0)
	return filepath.Join(filepath.Dir(file), "..", "README.md")
}()

// --- GitHub API types ---

type repo struct {
	Name            string    `json:"name"`
	Description     string    `json:"description"`
	HTMLURL         string    `json:"html_url"`
	StargazersCount int       `json:"stargazers_count"`
	ForksCount      int       `json:"forks_count"`
	Fork            bool      `json:"fork"`
	Private         bool      `json:"private"`
	UpdatedAt       time.Time `json:"updated_at"`
}

type event struct {
	Type    string      `json:"type"`
	Repo    eventRepo   `json:"repo"`
	Payload prPayload   `json:"payload"`
}

type eventRepo struct {
	Name string `json:"name"`
}

type prPayload struct {
	PullRequest prRef `json:"pull_request"`
}

type prRef struct {
	URL    string `json:"url"`
	Number int    `json:"number"`
}

type pullRequest struct {
	Title    string    `json:"title"`
	HTMLURL  string    `json:"html_url"`
	State    string    `json:"state"`
	MergedAt *string   `json:"merged_at"`
	Base     prBase    `json:"base"`
}

type prBase struct {
	Repo baseRepo `json:"repo"`
}

type baseRepo struct {
	Private bool `json:"private"`
}

// --- HTTP helper ---

func ghGet(path, token string, out any) error {
	req, err := http.NewRequest("GET", apiBase+path, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Accept", "application/vnd.github+json")
	req.Header.Set("X-GitHub-Api-Version", "2022-11-28")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	if resp.StatusCode >= 300 {
		return fmt.Errorf("GitHub API %s returned %d: %s", path, resp.StatusCode, body)
	}
	return json.Unmarshal(body, out)
}

// --- Section builders ---

func repoScore(r repo) float64 {
	daysAgo := time.Since(r.UpdatedAt).Hours() / 24
	recency := max(0, 365-daysAgo) / 365
	return float64(r.StargazersCount)*10 + recency
}

func buildProjectsSection(token string) (string, error) {
	var repos []repo
	if err := ghGet(fmt.Sprintf("/users/%s/repos?sort=updated&per_page=100&type=owner", githubUsername), token, &repos); err != nil {
		return "", err
	}

	var public []repo
	for _, r := range repos {
		if !r.Private && !r.Fork && r.Name != githubUsername {
			public = append(public, r)
		}
	}
	sort.Slice(public, func(i, j int) bool {
		return repoScore(public[i]) > repoScore(public[j])
	})
	if len(public) > topProjectsCount {
		public = public[:topProjectsCount]
	}

	var sb strings.Builder
	sb.WriteString("### What I've shipped lately\n\n")
	for _, r := range public {
		desc := strings.TrimRight(r.Description, ".")
		stars := ""
		if r.StargazersCount > 0 {
			stars = fmt.Sprintf(" ⭐ %d", r.StargazersCount)
		}
		fmt.Fprintf(&sb, "- **[%s](%s)**%s — %s.\n", r.Name, r.HTMLURL, stars, desc)
	}
	return strings.TrimRight(sb.String(), "\n"), nil
}

func buildOSSSection(token string) (string, error) {
	var events []event
	if err := ghGet(fmt.Sprintf("/users/%s/events?per_page=100", githubUsername), token, &events); err != nil {
		return "", err
	}

	type contribution struct {
		title, prURL, repoName, repoURL, status string
	}

	seen := map[string]bool{}
	var contribs []contribution

	for _, e := range events {
		if e.Type != "PullRequestEvent" {
			continue
		}
		parts := strings.SplitN(e.Repo.Name, "/", 2)
		if len(parts) < 2 || parts[0] == githubUsername {
			continue
		}

		apiURL := e.Payload.PullRequest.URL
		if apiURL == "" || seen[apiURL] {
			continue
		}
		seen[apiURL] = true

		// Fetch full PR to get title, state, and privacy flag
		var pr pullRequest
		apiPath := strings.TrimPrefix(apiURL, apiBase)
		if err := ghGet(apiPath, token, &pr); err != nil {
			continue
		}

		// Never show PRs in private repos
		if pr.Base.Repo.Private {
			continue
		}

		var status string
		switch {
		case pr.MergedAt != nil:
			status = "✅ Merged"
		case pr.State == "open":
			status = "🔄 Open"
		default:
			status = "❌ Closed"
		}

		prURL := pr.HTMLURL
		if prURL == "" {
			prURL = fmt.Sprintf("https://github.com/%s/pull/%d", e.Repo.Name, e.Payload.PullRequest.Number)
		}
		title := pr.Title
		if title == "" {
			title = e.Repo.Name
		}

		contribs = append(contribs, contribution{
			title:    title,
			prURL:    prURL,
			repoName: e.Repo.Name,
			repoURL:  "https://github.com/" + e.Repo.Name,
			status:   status,
		})
		if len(contribs) >= ossContribCount {
			break
		}
	}

	if len(contribs) == 0 {
		return "### Recent OSS\n\n_No recent external contributions found._", nil
	}

	var sb strings.Builder
	sb.WriteString("### Recent OSS\n\n")
	for _, c := range contribs {
		fmt.Fprintf(&sb, "- %s **[%s](%s)** into [%s](%s)\n", c.status, c.title, c.prURL, c.repoName, c.repoURL)
	}
	return strings.TrimRight(sb.String(), "\n"), nil
}

// --- Zone rewriter ---

func rewriteZone(content, zone, body string) (string, bool) {
	pattern := regexp.MustCompile(`(?s)(<!-- ` + zone + `_START -->)\n.*?\n(<!-- ` + zone + `_END -->)`)
	replaced := false
	result := pattern.ReplaceAllStringFunc(content, func(match string) string {
		replaced = true
		start := "<!-- " + zone + "_START -->"
		end := "<!-- " + zone + "_END -->"
		return start + "\n" + body + "\n" + end
	})
	return result, replaced
}

func max(a, b float64) float64 {
	if a > b {
		return a
	}
	return b
}

func main() {
	token := os.Getenv("GH_TOKEN")
	if token == "" {
		fmt.Fprintln(os.Stderr, "ERROR: GH_TOKEN environment variable is not set.")
		os.Exit(1)
	}

	absPath, err := filepath.Abs(readmePath)
	if err != nil {
		fmt.Fprintln(os.Stderr, "ERROR:", err)
		os.Exit(1)
	}

	original, err := os.ReadFile(absPath)
	if err != nil {
		fmt.Fprintln(os.Stderr, "ERROR reading README:", err)
		os.Exit(1)
	}

	projects, err := buildProjectsSection(token)
	if err != nil {
		fmt.Fprintln(os.Stderr, "ERROR fetching projects:", err)
		os.Exit(1)
	}

	oss, err := buildOSSSection(token)
	if err != nil {
		fmt.Fprintln(os.Stderr, "ERROR fetching OSS contributions:", err)
		os.Exit(1)
	}

	updated := string(original)
	updated, _ = rewriteZone(updated, "PROJECTS", projects)
	updated, _ = rewriteZone(updated, "OSS", oss)

	if updated == string(original) {
		fmt.Println("README.md is already up to date.")
		return
	}

	if err := os.WriteFile(absPath, []byte(updated), 0644); err != nil {
		fmt.Fprintln(os.Stderr, "ERROR writing README:", err)
		os.Exit(1)
	}
	fmt.Println("README.md updated.")
}
